source/extensions/common/wasm/wasm.cc

```diff
@@ -0,0 +1,455 @@
+
+const uint8_t* decodeVarint(const uint8_t* pos, const uint8_t* end, uint32_t* out) {
+  uint32_t ret = 0;
+  int shift = 0;
+  while (pos < end && (*pos & 0x80)) {
+    ret |= (*pos & 0x7f) << shift;
+    shift += 7;
+    pos++;
+  }
+  if (pos < end) {
+    ret |= *pos << shift;
+    pos++;
+  }
+  *out = ret;
+  return pos;
+}
+
+} // namespace
+
+Wasm::Wasm(absl::string_view runtime, absl::string_view vm_id, absl::string_view vm_configuration,
+           Stats::ScopeSharedPtr scope, Upstream::ClusterManager& cluster_manager,
+           Event::Dispatcher& dispatcher)
+    : vm_id_(std::string(vm_id)), wasm_vm_(Common::Wasm::createWasmVm(runtime, scope)),
+      scope_(scope), cluster_manager_(cluster_manager), dispatcher_(dispatcher),
+      time_source_(dispatcher.timeSource()), vm_configuration_(vm_configuration),
+      wasm_stats_(WasmStats{
+          ALL_WASM_STATS(POOL_COUNTER_PREFIX(*scope_, absl::StrCat("wasm.", runtime, ".")),
+                         POOL_GAUGE_PREFIX(*scope_, absl::StrCat("wasm.", runtime, ".")))}),
+      stat_name_set_(scope_->symbolTable().makeSet("Wasm").release()) {
+  active_wasm_++;
+  wasm_stats_.active_.set(active_wasm_);
+  wasm_stats_.created_.inc();
+  ENVOY_LOG(debug, "Base Wasm created {} now active", active_wasm_);
+}
+
+void Wasm::registerCallbacks() {
+#define _REGISTER(_fn)                                                                             \
+  wasm_vm_->registerCallback(                                                                      \
+      "env", #_fn, &Exports::_fn,                                                                  \
+      &ConvertFunctionWordToUint32<decltype(Exports::_fn),                                         \
+                                   Exports::_fn>::convertFunctionWordToUint32)
+  if (is_emscripten_) {
+    _REGISTER(pthread_equal);
+  }
+#undef _REGISTER
+
+#define _REGISTER_WASI(_fn)                                                                        \
+  wasm_vm_->registerCallback(                                                                      \
+      "wasi_unstable", #_fn, &Exports::wasi_unstable_##_fn,                                        \
+      &ConvertFunctionWordToUint32<decltype(Exports::wasi_unstable_##_fn),                         \
+                                   Exports::wasi_unstable_##_fn>::convertFunctionWordToUint32)
+  if (is_emscripten_) {
+    _REGISTER_WASI(fd_write);
+    _REGISTER_WASI(fd_seek);
+    _REGISTER_WASI(fd_close);
+    _REGISTER_WASI(environ_get);
+    _REGISTER_WASI(environ_sizes_get);
+    _REGISTER_WASI(proc_exit);
+  }
+#undef _REGISTER_WASI
+
+  // Calls with the "proxy_" prefix.
+#define _REGISTER_PROXY(_fn)                                                                       \
+  wasm_vm_->registerCallback(                                                                      \
+      "env", "proxy_" #_fn, &Exports::_fn,                                                         \
```

Comment:
Why do we register Envoy's ABI as part of `env` module rather than using a dedicated module name like `envoy` ?

Reply:
We originally had it like that, but ultimately decided against it.  As part of standardization we want to be WASI naming compatible, and using an "envoy" module would not fit with that.

Reply:
Can you link to more information whatever standardization process you are referring to? What is being standardized here? At a high level this sounds nice but I also want to make sure we are doing the right things here independent of whatever a standardizing body wants (not to mention I'm not sure we want to slow down to wait for a standardizing body).

Reply:
We discussed Proxy-WASM ABI design and standardization at the WebAssembly / WASI meeting on 10/15.  There was general interest in having a Proxy-WASM WASI standard and another similar proposal from Fastly.  The general feeling was that we should work together with the goal of standardization which would be in everyone's interest.  Of course standards take time and are best based on robust understanding of requirements which can often come only from implementation and production experience.  If follows that we need something which works well for all Envoy use-cases so we can get that experience.  However, given a choice of equals in the design it would be better to choose options which have fewer dependencies (i.e. not requiring protobuf support for basic things like configuration) which may not be available in other environments.

Reply:
And by not requiring protobufs for configuration, I mean in the WASM code in the argument to the on_configuration call.  In the host code in Envoy using protobufs is obviously the way to go.

Reply:
Per discussion elsewhere I really don't think we should be focusing on standardizing right now unless it provides very clear benefits. Everything seems very immature and we should focus now on doing the right thing ans satisfying real world use cases.